src/estimagic/optimization/subsolvers/_trsbox_quadratic_fast.py

```diff
@@ -0,0 +1,655 @@
+    x_candidate_new[x_bounded == -1] = lower_bounds[x_bounded == -1]
+    x_candidate_new[x_bounded == 1] = upper_bounds[x_bounded == 1]
+
+    return x_candidate_new
+
+
+@njit
+def _take_unconstrained_step_up_to_boundary(
+    raw_distance, gradient_sumsq, gradient_projected_sumsq, g_x, g_hess_g
+):
+    """Take unconstrained step, ignoring bounds, up to boundary."""
+    temp = np.sqrt(gradient_projected_sumsq * raw_distance + g_x**2)
+    if g_x >= 0:
+        distance_to_boundary = raw_distance / (temp + g_x)
+    else:
+        distance_to_boundary = (temp - g_x) / gradient_projected_sumsq
+    if g_hess_g <= 0:
+        step_len = distance_to_boundary[0]
+    else:
+        if distance_to_boundary <= gradient_sumsq / g_hess_g:
+            step_len = distance_to_boundary[0]
+        else:
+            step_len = gradient_sumsq / g_hess_g
+
+    return step_len, distance_to_boundary
+
+
+@njit
+def _update_candidate_vectors_and_reduction(
+    x_candidate,
+    x_bounded,
+    gradient_candidate,
+    gradient_projected,
+    step_len,
+    total_reduction,
+    curve_min,
+    index_bound_active,
+    gradient_projected_sumsq,
+    gradient_sumsq,
+    g_hess_g,
+    hess_g,
+):
+    """Update candidate vectors and the associated criterion reduction."""
+    current_min = g_hess_g / gradient_projected_sumsq
+
+    if index_bound_active.size == 0 and current_min > 0:
+        if curve_min != -1.0:
+            curve_min = min(curve_min, current_min)
+        else:
+            curve_min = current_min
+
+    gradient_sumsq_old = gradient_sumsq
+
+    gradient_candidate = gradient_candidate + step_len * hess_g
+    x_candidate = x_candidate + step_len * gradient_projected
+
+    gradient_sumsq = (
+        gradient_candidate[x_bounded == 0] @ gradient_candidate[x_bounded == 0]
+    )
+
+    current_reduction = max(
+        step_len * (gradient_sumsq_old - 0.5 * step_len * g_hess_g), 0
+    )
+    total_reduction = total_reduction + current_reduction
+
+    return (
+        x_candidate,
+        gradient_candidate,
+        current_reduction,
+        total_reduction,
+        curve_min,
+        gradient_sumsq,
+        gradient_sumsq_old,
+    )
+
+
+@njit
+def _take_constrained_step_up_to_boundary(
+    x_candidate, gradient_projected, step_len, lower_bounds, upper_bounds
+):
+    """Reduce step length, where boundary is hit, to preserve simple bounds."""
+    for i in range(len(x_candidate)):
+        if gradient_projected[i] != 0:
+            if gradient_projected[i] > 0:
+                step_len_constr = (
+                    upper_bounds[i] - x_candidate[i]
+                ) / gradient_projected[i]
+            else:
+                step_len_constr = (
+                    lower_bounds[i] - x_candidate[i]
+                ) / gradient_projected[i]
+            if step_len_constr < step_len:
+                step_len = step_len_constr
+                index_bound_active = np.array([i])
```

Comment:
Two questions:
1. is `index_bound_active` always defined? As far as I see it is only defined inside the if condition but who guarantees that this if condition is always true at least once?
2. Why is this an array of fixed length 1 instead of just an int?